Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -347,5 +382,151 @@ private bool TryInitializeCorrelationHelperIfNotInitialized()
+        /// </summary>
+        /// <remarks>
+        /// Unit tests should disable the ChildRequestTrackingSuppressionModule.
+        /// Unit test projects cannot create an [internal] IIS7WorkerRequest object.
+        /// Without this object, we cannot modify the Request.Headers without throwing a PlatformNotSupportedException.
+        /// Unit tests will have to initialize the RequestIdHeader.
+        /// The second IF will ensure the id is added to the activeRequests.
+        /// </remarks>
+        /// <remarks>
+        /// IIS Classic Pipeline should disable the ChildRequestTrackingSuppressionModule.
+        /// Classic does not create IIS7WorkerRequest object and Headers will be read-only.
+        /// (Exception System.PlatformNotSupportedException: This operation requires IIS integrated pipeline mode.)
+        /// </remarks>
+        private class ChildRequestTrackingSuppressionModule
+        {
+            private const int DEFAULTMAXVALUE = 100000;
+
+            private const string HeaderRootRequestId = "ApplicationInsights-RequestTrackingTelemetryModule-RootRequest-Id";
+#if DEBUG
+            private const string HeaderParentRequestId = "ApplicationInsights-RequestTrackingTelemetryModule-ParentRequest-Id";
+            private const string HeaderRequestId = "ApplicationInsights-RequestTrackingTelemetryModule-Request-Id";
+#endif
+
+            private static object semaphore = new object();
+
+            /// <summary>
+            /// Using this as a hash-set of current active requests. The second value is not used.
+            /// </summary>
+            private static ConcurrentDictionary<string, bool> activeRequestsA = new ConcurrentDictionary<string, bool>(System.Environment.ProcessorCount, DEFAULTMAXVALUE);
+            private static ConcurrentDictionary<string, bool> activeRequestsB = new ConcurrentDictionary<string, bool>(System.Environment.ProcessorCount, DEFAULTMAXVALUE);
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ChildRequestTrackingSuppressionModule" /> class.
+            /// </summary>
+            /// <param name="maxRequestsTracked">The maximum number of active requests to be tracked before resetting the dictionary.</param>
+            internal ChildRequestTrackingSuppressionModule(int maxRequestsTracked = DEFAULTMAXVALUE)
+            {
+                this.MAXSIZE = maxRequestsTracked > 0 ? maxRequestsTracked : DEFAULTMAXVALUE;
+            }
+
+            /// <summary>
+            /// Gets the Max number of request ids to cache.
+            /// </summary>
+            internal int MAXSIZE { get; private set; }
+
+            /// <summary>
+            /// Request will be tagged with an id to identify if it should be logged later.
+            /// </summary>
+            internal void OnBeginRequest_IdRequest(HttpContext context)
+            {
+                this.TagRequest(context);
```

Comment:
both begin and end methods in this module should be try..catch..log.. to be safe